File: src/class036/C2_ZigzagLevelOrderTraversal.py
# 二叉树的锯齿形层序遍历
# 测试链接 : https://leetcode.cn/problems/binary-tree-zigzag-level-order-traversal/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 提交以下的方法
# 用每次处理一层的优化bfs就非常容易实现
# 如果测试数据量变大了就修改这个值
MAXN = 2001

queue = [None] * MAXN
l, r = 0, 0

def zigzagLevelOrder(root):
    ans = []
    if root is not None:
        global l, r
        l, r = 0, 0
        queue[r] = root
        r += 1
        # false 代表从左往右
        # true 代表从右往左
        reverse = False
        while l < r:
            size = r - l
            list_ = []
            # reverse == False, 左 -> 右, l....r-1, 收集size个
            # reverse == True, 右 -> 左, r-1....l, 收集size个
            # 左 -> 右, i = i + 1
            # 右 -> 左, i = i - 1
            for i in range(r-1, l-1, -1) if reverse else range(l, r):
                cur = queue[i]
                list_.append(cur.val)
            for i in range(size):
                cur = queue[l]
                l += 1
                if cur.left is not None:
                    queue[r] = cur.left
                    r += 1
                if cur.right is not None:
                    queue[r] = cur.right
                    r += 1
            ans.append(list_)
            reverse = not reverse
    return ans

File: src/class036/C2_ZigzagLevelOrderTraversal.cpp
#include <vector>
#include <deque>

using namespace std;

// 二叉树的锯齿形层序遍历
// 测试链接 : https://leetcode.cn/problems/binary-tree-zigzag-level-order-traversal/
class Solution {
public:
    struct TreeNode {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    };

    // 用每次处理一层的优化bfs就非常容易实现
    // 如果测试数据量变大了就修改这个值
    static const int MAXN = 2001;
    TreeNode* queue[MAXN];
    int l, r;

    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (root != nullptr) {
            l = r = 0;
            queue[r++] = root;
            // false 代表从左往右
            // true 代表从右往左
            bool reverse = false; 
            while (l < r) {
                int size = r - l;
                vector<int> list;
                // reverse == false, 左 -> 右， l....r-1, 收集size个
                // reverse == true, 右 -> 左， r-1....l, 收集size个
                // 左 -> 右, i = i + 1
                // 右 -> 左, i = i - 1
                int start = reverse ? r - 1 : l;  
                int step = reverse ? -1 : 1;              
                // 从 start 开始，根据 step 的正负值递增或递减 i  
                for (int i = start, k = 0; k < size; i += step, k++) { 
                    TreeNode* cur = queue[i];
                    list.push_back(cur->val);
                }
                for (int i = 0; i < size; i++) {
                    TreeNode* cur = queue[l++];
                    if (cur->left != nullptr) {
                        queue[r++] = cur->left;
                    }
                    if (cur->right != nullptr) {
                        queue[r++] = cur->right;
                    }
                }
                ans.push_back(list);
                reverse = !reverse;
            }
        }
        return ans;
    }
};
